src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1446,8 +1477,24 @@ RouterJob(Query *originalQuery, RelationRestrictionContext *restrictionContext,
 	}
 	else if (isMultiShardModifyQuery)
 	{
-		job->taskList = MultiShardModifyTaskList(originalQuery, relationShardList,
+		job->taskList = QueryPushdownSqlTaskList(originalQuery, 0,
+												 plannerRestrictionContext->
+												 relationRestrictionContext,
+												 relationShardList, MODIFY_TASK,
 												 requiresMasterEvaluation);
+
+		/* If it is a modify task with multiple tables */
+		if (list_length(
+				plannerRestrictionContext->relationRestrictionContext->
+				relationRestrictionList) > 1)
+		{
+			ListCell *taskCell = NULL;
+			foreach(taskCell, job->taskList)
+			{
+				Task *task = (Task *) lfirst(taskCell);
```

Comment:
Can't we move this logic into `QueryPushdownSqlTaskList ()` and apply this when the task is `ModifyTask`? That seems a better place, no?